File: apps/js/src/routes/me.ts
import { Router } from "express";

import { fail } from "#/lib/problem";
import { wire } from "#/lib/wire";
import { auth } from "#/middleware/auth";
import * as users from "#/service/user";

export const router: Router = Router();

/**
 * @openapi
 * components:
 *   schemas:
 *     User:
 *       type: object
 *       properties:
 *         id: { type: string }
 *         email: { type: string }
 *         created_at: { type: string }
 *         updated_at: { type: string }
 *         name: { type: string }
 *         phone: { type: string }
 *         birthdate: { type: string }
 *         gender: { type: string }
 *         avatar: { type: string }
 *         roles:
 *           type: array
 *           items: { type: string }
 *           uniqueItems: false
 *       required: [created_at, email, id, roles, updated_at]
 *     UserPayment:
 *       type: object
 *       properties:
 *         id: { type: string }
 *         created_at: { type: string }
 *         id_payment: { type: string }
 *         type: { type: string }
 *         is_default: { type: boolean }
 *         data: { type: object }
 *       required: [created_at, data, id, id_payment, is_default, type]
 *
 * /me:
 *   get:
 *     summary: Fetch the profile
 *     tags: [me]
 *     security: [{ BearerAuth: [] }]
 *     responses:
 *       "200":
 *         description: OK
 *         content:
 *           application/json:
 *             schema: { $ref: "#/components/schemas/User" }
 *       "401":
 *         description: Invalid token
 *         content:
 *           application/json:
 *             schema: { $ref: "#/components/schemas/Problem" }
 *       "404":
 *         description: No such user
 *         content:
 *           application/json:
 *             schema: { $ref: "#/components/schemas/Problem" }
 *       "500":
 *         description: Internal error
 *         content:
 *           application/json:
 *             schema: { $ref: "#/components/schemas/Problem" }
 */
router.get("/me", auth, async (req, res) => {
	try {
		res.json(wire(await users.me(req.idUser)));
	} catch (error) {
		fail(res, error);
	}
});

/**
 * @openapi
 * /me/payments:
 *   get:
 *     summary: List saved payment methods
 *     tags: [me]
 *     security: [{ BearerAuth: [] }]
 *     responses:
 *       "200":
 *         description: OK
 *         content:
 *           application/json:
 *             schema:
 *               { type: array, items: { $ref: "#/components/schemas/UserPayment" } }
 *       "401":
 *         description: Invalid token
 *         content:
 *           application/json:
 *             schema: { $ref: "#/components/schemas/Problem" }
 *       "500":
 *         description: Internal error
 *         content:
 *           application/json:
 *             schema: { $ref: "#/components/schemas/Problem" }
 */
// TODO: add POST /me/payments; a fresh fixture has none to validate
router.get("/me/payments", auth, async (req, res) => {
	try {
		res.json(wire(await users.payments(req.idUser)));
	} catch (error) {
		fail(res, error);
	}
});

File: apps/go/internal/handler/me.go
package handler

import (
	"errors"
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/ryansuhartanto/koda-b8-backend/apps/go/internal/middleware"
	"github.com/ryansuhartanto/koda-b8-backend/apps/go/internal/model"
	"github.com/ryansuhartanto/koda-b8-backend/apps/go/internal/repository"
)

func Me(r *gin.Engine, pool *pgxpool.Pool) {
	r.GET("/me", middleware.Auth(), me(pool))
	r.GET("/me/payments", middleware.Auth(), listPayments(pool))
}

// @Summary  Fetch the caller's profile
// @Tags     me
// @Produce  json
// @Security BearerAuth
// @Success  200 {object} model.UsersMe    "OK"
// @Failure  401 {object} model.Problem "Missing or invalid token"
// @Failure  404 {object} model.Problem "No such user"
// @Failure  500 {object} model.Problem "Internal error"
// @Router   /me [get]
func me(pool *pgxpool.Pool) gin.HandlerFunc {
	return func(ctx *gin.Context) {
		user, err := repository.Me(ctx, pool, ctx.GetInt64(middleware.ContextIDUser))
		if err != nil {
			// the token outlived the account
			if errors.Is(err, pgx.ErrNoRows) {
				model.AbortProblem(ctx, http.StatusNotFound, "no such user")
				return
			}

			model.AbortProblem(ctx, http.StatusInternalServerError, err.Error())
			return
		}

		ctx.PureJSON(http.StatusOK, user)
	}
}

// @Summary  List the caller's saved payment methods
// @Tags     me
// @Produce  json
// @Security BearerAuth
// @Success  200 {array}  model.UsersPaymentsActive "OK"
// @Failure  401 {object} model.Problem     "Missing or invalid token"
// @Failure  500 {object} model.Problem     "Internal error"
// @Router   /me/payments [get]
// TODO: add POST /me/payments; a fresh fixture has none to validate
func listPayments(pool *pgxpool.Pool) gin.HandlerFunc {
	return func(ctx *gin.Context) {
		payments, err := repository.Payments(ctx, pool, ctx.GetInt64(middleware.ContextIDUser))
		if err != nil {
			model.AbortProblem(ctx, http.StatusInternalServerError, err.Error())
			return
		}

		ctx.PureJSON(http.StatusOK, payments)
	}
}
